src/main/scala/streamit/client/ApacheKafkaClient.scala

```diff
@@ -0,0 +1,257 @@
+        )(
+          (ac, _) =>
+            ConcurrentEffect[F].delay(ac.close()) *> logger.info(s"Zookeeper client Closed")
+        )
+    }
+
+  /**
+    * @return A [[fs2.Stream]] containing an [[org.apache.kafka.clients.admin.AdminClient]], which will be
+    *   auto-closed via Stream.bracket.
+    */
+  def createAdminClient(): Stream[F, AdminClient] = {
+    val properties = new Properties()
+    properties.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, settings.bootstrapServers)
+    Stream.bracketCase(
+      ConcurrentEffect[F].delay(AdminClient.create(properties)) <*
+        logger.info(
+          s"Created admin client to talk to bootstrap server: ${settings.bootstrapServers}"
+        )
+    )((ac, _) => ConcurrentEffect[F].delay(ac.close()) *> logger.info(s"Admin client closed"))
+  }
+
+  /**
+    * @return A [[fs2.Stream]] containing an [[org.apache.kafka.clients.consumer.KafkaConsumer]], which will be
+    *   auto-closed via Stream.bracket.
+    */
+  def createConsumer[K, V](
+    pollSize: Option[Int] = None,
+    bootstrapServers: Option[BootstrapServers] = None
+  )(implicit keyDes: Deserializer[K], valDes: Deserializer[V]): Stream[F, KafkaConsumer[K, V]] = {
+
+    val initialProps = new Properties()
+    initialProps.put(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG, "30000")
+    initialProps.put(ConsumerConfig.CLIENT_ID_CONFIG, KafkaClientId)
+    initialProps.put(ConsumerConfig.GROUP_ID_CONFIG, settings.groupId.toString)
+    initialProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, keyDes.getClass)
+    initialProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valDes.getClass)
+    pollSize.map(f => initialProps.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, f.toString))
+    initialProps
+      .put(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG, settings.schemaRegistry)
+
+    Stream(initialProps)
+      .map(props => {
+        props.put(
+          ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+          bootstrapServers.getOrElse(settings.bootstrapServers)
+        )
+        props
+      })
+      .flatMap(props => {
+        Stream.bracketCase(
+          ConcurrentEffect[F].delay(new KafkaConsumer[K, V](props)) <*
+            logger.info(
+              s"Created consumer to bootstrap server: ${settings.bootstrapServers}, listing topics"
+            )
+        )((c, _) => ConcurrentEffect[F].delay(c.close()) *> logger.info("Consumer closed"))
+      })
+  }
+
+  /**
+    * @return A [[fs2.Stream]] containing an [[org.apache.kafka.clients.consumer.KafkaConsumer]], which will be
+    *   auto-closed via Stream.bracket.
+    */
+  def createConsumerForTopic[K, V](
+    topic: Topic,
+    offset: Option[Duration] = None,
+    pollSize: Option[Int] = None
+  )(implicit keyDes: Deserializer[K], valDes: Deserializer[V]): Stream[F, KafkaConsumer[K, V]] = {
+
+    def resetOffsetListener(
+      desc: String,
+      offsetChangeFunc: JavaCollection[TopicPartition] => Unit,
```

Comment:
the function `offsetChangeFunc` is side effecting. It's probably related to the kafka client library. It can be interesting to see how far we can go describing it as `JavaCollection[TopicPartition] => F[Unit]`. May be in another change set as said above.

Reply:
Yep - the apache kafka API doesn't really give much flexibility, am happy if you want to have a shot at it :) (yes, in another PR please:))